client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkUtils.java

```diff
@@ -92,6 +95,26 @@ public static String appUniqueId(SparkContext context) {
         .getOrElse(context::applicationId);
   }
 
+  public static int celebornShuffleId(
+      ShuffleClient client,
+      CelebornShuffleHandle<?, ?, ?> handle,
+      TaskContext context,
+      Boolean isWriter) {
+    if (handle.throwsFetchFailure()) {
+      String appShuffleIdentifier =
+          new StringBuilder()
+              .append(handle.shuffleId())
+              .append("-")
+              .append(context.stageId())
+              .append("-")
+              .append(context.stageAttemptNumber())
+              .toString();
+      return client.getShuffleId(handle.shuffleId(), appShuffleIdentifier, isWriter);
```

Comment:
QQ: Do we want to cache this ? It should be immutable for a given shuffle id + stage id + attempt across all tasks.

Reply:
+1

Reply:
caching shuffleid is a pretty good idea, however, I found it hard to implement in shuffleIdTracker.
shuffleIdTracker locates in SparkShuffleManager, which is not always true to have SparkShuffleManager when shuffleid is needed, for example

```
public class HashBasedShuffleWriterSuiteJ extends CelebornShuffleWriterSuiteBase {

  @Override
  protected ShuffleWriter<Integer, String> createShuffleWriter(
      CelebornShuffleHandle handle,
      TaskContext context,
      CelebornConf conf,
      ShuffleClient client,
      ShuffleWriteMetricsReporter metrics)
      throws IOException {
    return new HashBasedShuffleWriter<Integer, String, String>(
        SparkUtils.celebornShuffleId(client, handle, context, true),
        handle,
        context,
        conf,
        client,
        metrics,
        SendBufferPool.get(1, 30, 60));
  }
}
```

Currently, `shuffleIdCache` is added to ShuffleClientImpl, and works very similar to `reducePartitionMap` 
```
  // key: appShuffleIdentifier, value: shuffleId
  protected Map<String, Integer> shuffleIdCache = JavaUtils.newConcurrentHashMap();
```